### PR TITLE
Parse auth challenges where Bearer is not the first challenge. #1042

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 1]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+    else
+    {
+        NSString *bearerChallendge = [self extractBearerChallenge:headerContents];
+        if (!bearerChallendge) {
+            adError = [self invalidHeader:headerContents];
+        } else {
+            parameters = [self parseBearerChallenge:bearerChallendge error:&adError];
+        }
+    }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 2]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        } else {
+            parameters = [self parseBearerChallenge:bearerChallendge error:&adError];
+        }
+    }
+    
+    if (error)
+    {
+        *error = adError;
+    }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 3]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        
+        if (c == ',')
+        {
+            if (possibleEndOfChallenge)
+            {
+                // Next challenge found.
+                possibleEndOfChallenge = NO;
+                break;
+            }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 4]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+                // Next challenge found.
+                possibleEndOfChallenge = NO;
+                break;
+            }
+            
+            possibleEndOfChallenge = YES;
+            nonSpaceOrEqualCharacterDetected = NO;
+            spaceCharacterDetected = NO;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 5]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+    
+    return parameters;
+}
+
+#pragma mark - Private
+
++ (NSUInteger)quotedStringLastIndex:(NSString *)headerContents startIndex:(NSUInteger)startIndex
+{
+    NSUInteger lastIndex = startIndex;
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 6]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+            possibleEndOfChallenge = YES;
+            nonSpaceOrEqualCharacterDetected = NO;
+            spaceCharacterDetected = NO;
+        }
+        
+        if (!possibleEndOfChallenge)
+        {
+            lastChallengeIndex = nextIndex;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 7]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+{
+    NSUInteger lastIndex = startIndex;
+    
+    unichar quotedChar = [headerContents characterAtIndex:startIndex];
+    if (quotedChar == '"' || quotedChar == '\'')
+    {
+        NSUInteger nextIndex = startIndex + 1;
+        while (nextIndex < headerContents.length) {
+            unichar c = [headerContents characterAtIndex:nextIndex];
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 8]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        if (!possibleEndOfChallenge)
+        {
+            lastChallengeIndex = nextIndex;
+        }
+        
+        if (c != ' ' && c != ',')
+        {
+            nonSpaceOrEqualCharacterDetected = YES;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 9]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+    {
+        NSUInteger nextIndex = startIndex + 1;
+        while (nextIndex < headerContents.length) {
+            unichar c = [headerContents characterAtIndex:nextIndex];
+            
+            if (c == quotedChar) {
+                lastIndex = nextIndex;
+                break;
+            }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 10]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        if (c != ' ' && c != ',')
+        {
+            nonSpaceOrEqualCharacterDetected = YES;
+        }
+        
+        if (possibleEndOfChallenge && nonSpaceOrEqualCharacterDetected && c == ' ')
+        {
+            spaceCharacterDetected = YES;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 11]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+            
+            if (c == quotedChar) {
+                lastIndex = nextIndex;
+                break;
+            }
+            
+            nextIndex++;
+        }
+    }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 12]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+            }
+            
+            nextIndex++;
+        }
+    }
+    
+    return lastIndex;
+}
+
++ (NSString *)extractBearerChallenge:(NSString *)headerContents
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: CRITICAL Function has a complexity of 23 which is greater than 20 authorized. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AFunctionComplexity)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 13]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        
+        if (possibleEndOfChallenge && nonSpaceOrEqualCharacterDetected && c == '=')
+        {
+            // Next parameter found.
+            possibleEndOfChallenge = NO;
+            nonSpaceOrEqualCharacterDetected = NO;
+            spaceCharacterDetected = NO;
+            lastChallengeIndex = nextIndex;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 14]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        if (possibleEndOfChallenge && nonSpaceOrEqualCharacterDetected && c == '=')
+        {
+            // Next parameter found.
+            possibleEndOfChallenge = NO;
+            nonSpaceOrEqualCharacterDetected = NO;
+            spaceCharacterDetected = NO;
+            lastChallengeIndex = nextIndex;
+        }
+        
+        if (possibleEndOfChallenge && nonSpaceOrEqualCharacterDetected && spaceCharacterDetected && c != ',' && c != ' ' && c != ' ')
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MAJOR Identical sub-expressions on both sides of operator "&&" [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1764)
MAJOR Split this 133 characters long line (which is greater than 120 authorized). [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3ALineLength)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 15]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        }
+    }
+    
+    return lastIndex;
+}
+
++ (NSString *)extractBearerChallenge:(NSString *)headerContents
+{
+    NSRange range = [headerContents rangeOfString:@"Bearer "];
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 16]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
++ (NSString *)extractBearerChallenge:(NSString *)headerContents
+{
+    NSRange range = [headerContents rangeOfString:@"Bearer "];
+    
+    if (range.location == NSNotFound)
+    {
+        return nil;
+    }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 17]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+            lastChallengeIndex = nextIndex;
+        }
+        
+        if (possibleEndOfChallenge && nonSpaceOrEqualCharacterDetected && spaceCharacterDetected && c != ',' && c != ' ' && c != ' ')
+        {
+            // Next challenge found.
+            possibleEndOfChallenge = NO;
+            break;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 18]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+    NSRange range = [headerContents rangeOfString:@"Bearer "];
+    
+    if (range.location == NSNotFound)
+    {
+        return nil;
+    }
+    
+    NSUInteger nextIndex = range.location + range.length;
+    NSUInteger lastChallengeIndex = NSNotFound;
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 19]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        if (possibleEndOfChallenge && nonSpaceOrEqualCharacterDetected && spaceCharacterDetected && c != ',' && c != ' ' && c != ' ')
+        {
+            // Next challenge found.
+            possibleEndOfChallenge = NO;
+            break;
+        }
+        
+        nextIndex++;
+    }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 20]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        return nil;
+    }
+    
+    NSUInteger nextIndex = range.location + range.length;
+    NSUInteger lastChallengeIndex = NSNotFound;
+    
+    BOOL possibleEndOfChallenge = NO;
+    BOOL nonSpaceOrEqualCharacterDetected = NO;
+    BOOL spaceCharacterDetected = NO;
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 21]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+    }
+    
+    NSUInteger nextIndex = range.location + range.length;
+    NSUInteger lastChallengeIndex = NSNotFound;
+    
+    BOOL possibleEndOfChallenge = NO;
+    BOOL nonSpaceOrEqualCharacterDetected = NO;
+    BOOL spaceCharacterDetected = NO;
+    
+    while (nextIndex < headerContents.length)
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MAJOR Reduce the number of nested break and goto statements from 2 to 1 authorized. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3ASingleGotoOrBreakPerIteration)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 22]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
     return nil;
+
+}
+
++ (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
+                                       error:(ADAuthenticationError * __autoreleasing *)error;
+{
+    __block ADAuthenticationError* adError = nil;
+    NSDictionary *parameters = nil;
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 23]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+    NSUInteger lastChallengeIndex = NSNotFound;
+    
+    BOOL possibleEndOfChallenge = NO;
+    BOOL nonSpaceOrEqualCharacterDetected = NO;
+    BOOL spaceCharacterDetected = NO;
+    
+    while (nextIndex < headerContents.length)
+    {
+        unichar c = [headerContents characterAtIndex:nextIndex];
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 24]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        nextIndex++;
+    }
+    
+    if (possibleEndOfChallenge)
+    {
+        // There is a ',' at the end of the string that is not followed by a param or a chellenge.
+        // That is not allowed.
+        return nil;
+    }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 25]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+    }
+    
+    if (possibleEndOfChallenge)
+    {
+        // There is a ',' at the end of the string that is not followed by a param or a chellenge.
+        // That is not allowed.
+        return nil;
+    }
+    
+    NSUInteger length = lastChallengeIndex + 1 - range.location;
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Add parentheses to make the operator precedence explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS864)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 26]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+    NSString *result = [headerContents substringWithRange:NSMakeRange(range.location, length)];
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 27]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        
+        // Start of a quaoted string, lets get last index of it and continue iteration.
+        if (c == '"' || c == '\'')
+        {
+            nextIndex = [self quotedStringLastIndex:headerContents startIndex:nextIndex];
+            lastChallengeIndex = nextIndex;
+            nextIndex++;
+            continue;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

**Comment by partnerinflight** [comment]
SonarQube analysis reported 28 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 24 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 28]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+                // Next challenge found.
+                possibleEndOfChallenge = NO;
+                break;
+            }
+            
+            possibleEndOfChallenge = YES;
+            paramOrTokenNameDetected = NO;
+            spaceCharacterDetected = NO;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 29]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        if (!possibleEndOfChallenge)
+        {
+            lastChallengeIndex = nextIndex;
+        }
+        
+        if (c != ' ' && c != ',' && c != '=')
+        {
+            paramOrTokenNameDetected = YES;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 30]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        if (c != ' ' && c != ',' && c != '=')
+        {
+            paramOrTokenNameDetected = YES;
+        }
+        
+        if (possibleEndOfChallenge && paramOrTokenNameDetected && c == ' ')
+        {
+            spaceCharacterDetected = YES;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 31]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        
+        if (possibleEndOfChallenge && paramOrTokenNameDetected && c == '=')
+        {
+            // Next parameter found.
+            possibleEndOfChallenge = NO;
+            paramOrTokenNameDetected = NO;
+            spaceCharacterDetected = NO;
+            lastChallengeIndex = nextIndex;
+        }
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 32]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        if (possibleEndOfChallenge && paramOrTokenNameDetected && c == '=')
+        {
+            // Next parameter found.
+            possibleEndOfChallenge = NO;
+            paramOrTokenNameDetected = NO;
+            spaceCharacterDetected = NO;
+            lastChallengeIndex = nextIndex;
+        }
+        
+        if (possibleEndOfChallenge && paramOrTokenNameDetected && spaceCharacterDetected && c != ',' && c != ' ' && c != '=')
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MAJOR Split this 125 characters long line (which is greater than 120 authorized). [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3ALineLength)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 33]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+        return nil;
+    }
+    
+    NSUInteger nextIndex = range.location + range.length;
+    NSUInteger lastChallengeIndex = NSNotFound;
+    
+    BOOL possibleEndOfChallenge = NO;
+    BOOL paramOrTokenNameDetected = NO;
+    BOOL spaceCharacterDetected = NO;
+
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1131)

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 34]
```diff
@@ -168,6 +168,150 @@ + (NSDictionary *)extractChallengeParameters:(NSString *)headerContents
...
+    }
+    
+    NSUInteger nextIndex = range.location + range.length;
+    NSUInteger lastChallengeIndex = NSNotFound;
+    
+    BOOL possibleEndOfChallenge = NO;
+    BOOL paramOrTokenNameDetected = NO;
+    BOOL spaceCharacterDetected = NO;
+    
+    while (nextIndex < headerContents.length)
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MAJOR Reduce the number of nested break and goto statements from 2 to 1 authorized. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3ASingleGotoOrBreakPerIteration)

**Comment by partnerinflight** [comment]
SonarQube analysis reported 27 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 24 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 35]
```diff
@@ -95,79 +96,175 @@ + (ADAuthenticationError *)invalidHeader:(NSString *)headerContents
...
+        NSString *item = items[i];
+        NSRange range = [item rangeOfString:BearerWithParamExpression options:NSRegularExpressionSearch];
+        if (range.location != NSNotFound)
+        {
+            // Remove 'Bearer'.
+            NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:BearerSchemaNameExpression options:NSRegularExpressionCaseInsensitive error:nil];
+            NSString *param = [regex stringByReplacingMatchesInString:item options:0 range:NSMakeRange(0, [item length]) withTemplate:@""];
+            
+            // Save cleared param.
+            items[i] = param;
```
Comment (partnerinflight) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: MAJOR Remove this hazardous cast. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS856)

**Comment by partnerinflight** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 36]
```diff
@@ -95,79 +96,175 @@ + (ADAuthenticationError *)invalidHeader:(NSString *)headerContents
...
     }
     
-    if (error)
+    return lastIndex;
+}
+
++ (BOOL)isParameter:(NSString *)string
+{
+    NSRange range = [string rangeOfString:ExtractionExpression options:NSRegularExpressionSearch];
+    BOOL result = range.location == 0;
```
Comment (RPangrle) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: Any reason to not just return here?

Comment (antrix1989) [reply ADAL/src/ADAuthenticationParameters+Internal.m]: I used it for debugging and forgot to remove. Will replace it with return.

File: ADAL/src/ADAuthenticationParameters+Internal.m [hunk 37]
```diff
@@ -38,8 +38,9 @@
 NSString* const UnauthorizedHTTStatusExpected = @"Unauthorized (401) HTTP status code is expected, but the actual status code is %d";
 const unichar Quote = '\"';
 //The regular expression that matches the Bearer contents:
-NSString* const RegularExpression = @"^Bearer\\s+([^,\\s=\"]+?)=\"([^\"]*?)\"\\s*(?:,\\s*([^,\\s=\"]+?)=\"([^\"]*?)\"\\s*)*$";
-NSString* const ExtractionExpression = @"\\s*([^,\\s=\"]+?)=\"([^\"]*?)\"";
+NSString* const BearerWithParamExpression = @"\\s*Bearer\\s+([^,\\s=\"]+?)\\s*=\\s*\"([^\"]*?)\"";
```
Comment (RPangrle) [inline ADAL/src/ADAuthenticationParameters+Internal.m]: My preference is for regex to be defined where it's used, and if a regex expression is used in more then one place that gets moved into a helper method.

Comment (antrix1989) [reply ADAL/src/ADAuthenticationParameters+Internal.m]: Removed all but ExtractionExpression. The code will be more complicated if we move it into a helper method.

**Pull request merged**